File: agents/sim/agent.mjs
#!/usr/bin/env node
// OpenGym turnike simülatörü — Node >= 22 (global WebSocket), sıfır bağımlılık.
//
// Kullanım:
//   DEVICE_ID=... DEVICE_TOKEN=og_... node agent.mjs
//
// Cihaz artık "dumb client": kimlik doğrular, bağlı kalır ve sunucudan gelen
// "open" komutunu bekler (üye taraması artık cihazda değil, telefon
// uygulamasında gerçekleşir). Bu ajan yalnızca bağlantı/röle simülasyonudur.
//
// Ortam değişkenleri:
//   GATEWAY_URL   varsayılan: ws://127.0.0.1:3000/api/device-gateway
//   DEVICE_ID     panelden eklenen cihazın id'si
//   DEVICE_TOKEN  cihaz eklenirken yalnızca bir kez gösterilen "og_" önekli token

import process from "node:process";

const GATEWAY_URL =
  process.env.GATEWAY_URL ?? "ws://127.0.0.1:3000/api/device-gateway";
const DEVICE_ID = process.env.DEVICE_ID;
const DEVICE_TOKEN = process.env.DEVICE_TOKEN;

if (!DEVICE_ID || !DEVICE_TOKEN) {
  console.error("DEVICE_ID ve DEVICE_TOKEN ortam değişkenleri zorunlu.");
  process.exit(2);
}

let reconnectDelayMs = 1000;

function connect() {
  const ws = new WebSocket(GATEWAY_URL);

  ws.addEventListener("open", () => {
    ws.send(
      JSON.stringify({ type: "auth", deviceId: DEVICE_ID, token: DEVICE_TOKEN }),
    );
  });

  ws.addEventListener("message", (event) => {
    let msg;
    try {
      msg = JSON.parse(String(event.data));
    } catch {
      return;
    }

    if (msg.type === "auth_ok") {
      reconnectDelayMs = 1000;
      console.log(`[bağlı] cihaz: ${msg.deviceName} — açılma komutu bekleniyor…`);
      return;
    }

    if (msg.type === "auth_error") {
      // Token yanlışsa yeniden denemenin anlamı yok — çık.
      console.error(`[kimlik hatası] ${msg.message}`);
      process.exit(2);
    }

    if (msg.type === "open") {
      console.log(`AÇIK — röle ${msg.openMs ?? 500} ms tetiklendi`);
    }
  });

  ws.addEventListener("close", () => {
    console.log(
      `[koptu] ${Math.round(reconnectDelayMs / 1000)} sn sonra yeniden bağlanılacak...`,
    );
    setTimeout(connect, reconnectDelayMs);
    reconnectDelayMs = Math.min(reconnectDelayMs * 2, 30_000);
  });

  // "error" sonrasında her zaman "close" gelir; yeniden bağlanmayı close yönetir.
  ws.addEventListener("error", () => {});
}

connect();

File: agents/rpi/agent.py
#!/usr/bin/env python3
# OpenGym turnike ajanı — Raspberry Pi referans uygulaması.
#
# Gereksinimler:
#   pip install websockets gpiozero
#
# Donanım varsayımı:
#   - Röle modülü GPIO pinine bağlı (varsayılan BCM 17), turnike tetik girişini sürer.
#   - QR taraması artık cihazda değil, üyenin telefon uygulamasında yapılır;
#     bu ajan yalnızca kimlik doğrular ve sunucudan gelen "open" komutuyla röleyi tetikler.
#
# Ortam değişkenleri:
#   GATEWAY_URL   varsayılan: ws://127.0.0.1:3000/api/device-gateway
#   DEVICE_ID     panelden eklenen cihazın id'si
#   DEVICE_TOKEN  cihaz eklenirken yalnızca bir kez gösterilen "og_" önekli token
#   RELAY_GPIO    varsayılan: 17 (BCM)
#
# Fail-closed: röle varsayılan olarak KAPALI; yalnızca sunucudan "open" geldiğinde
# openMs süresince açılır. Bağlantı yokken hiçbir komut kapı açmaz.

import asyncio
import json
import os
import sys

try:
    import websockets
except ImportError:
    sys.exit("websockets paketi gerekli: pip install websockets")

GATEWAY_URL = os.environ.get("GATEWAY_URL", "ws://127.0.0.1:3000/api/device-gateway")
DEVICE_ID = os.environ.get("DEVICE_ID")
DEVICE_TOKEN = os.environ.get("DEVICE_TOKEN")
RELAY_GPIO = int(os.environ.get("RELAY_GPIO", "17"))

if not DEVICE_ID or not DEVICE_TOKEN:
    sys.exit("DEVICE_ID ve DEVICE_TOKEN ortam değişkenleri zorunlu.")


class DummyRelay:
    """GPIO olmayan ortamda (geliştirme makinesi) simülasyon."""

    def on(self):
        print("[röle] AÇIK (simülasyon)")

    def off(self):
        print("[röle] KAPALI (simülasyon)")


try:
    from gpiozero import OutputDevice

    # initial_value=False → açılışta röle kapalı (fail-closed)
    relay = OutputDevice(RELAY_GPIO, active_high=True, initial_value=False)
except Exception:
    relay = DummyRelay()


async def pulse_relay(open_ms: int) -> None:
    # Röle yalnızca süreli açılır; her durumda finally ile kapanır.
    relay.on()
    try:
        await asyncio.sleep(open_ms / 1000)
    finally:
        relay.off()


async def handle_messages(ws) -> None:
    async for raw in ws:
        try:
            msg = json.loads(raw)
        except json.JSONDecodeError:
            continue
        if msg.get("type") == "open":
            print("AÇIK")
            asyncio.create_task(pulse_relay(int(msg.get("openMs", 500))))


async def run() -> None:
    delay = 1
    while True:
        try:
            # websockets kütüphanesi sunucu ping'lerini otomatik yanıtlar (keepalive).
            async with websockets.connect(GATEWAY_URL) as ws:
                await ws.send(
                    json.dumps(
                        {"type": "auth", "deviceId": DEVICE_ID, "token": DEVICE_TOKEN}
                    )
                )
                first = json.loads(await ws.recv())
                if first.get("type") != "auth_ok":
                    # Token yanlışsa yeniden denemenin anlamı yok — çık.
                    sys.exit(f"[kimlik hatası] {first.get('message', '?')}")
                print(f"[bağlı] cihaz: {first.get('deviceName')} — açılma komutu bekleniyor…")
                delay = 1

                await handle_messages(ws)  # bağlantı kopunca döner
        except (OSError, websockets.WebSocketException) as exc:
            print(f"[koptu] {exc} — {delay} sn sonra yeniden bağlanılacak...")

        await asyncio.sleep(delay)
        delay = min(delay * 2, 30)


if __name__ == "__main__":
    try:
        asyncio.run(run())
    except KeyboardInterrupt:
        pass
    finally:
        relay.off()  # her çıkışta fail-closed
